apps/backend/src/build/finalizeBuild.ts

```diff
@@ -1,18 +1,122 @@
 import { TransactionOrKnex } from "objection";
 
-import { Build, Screenshot } from "@/database/models/index.js";
+import { Build, BuildShard, Screenshot } from "@/database/models/index.js";
+import { BuildMetadata } from "@/database/schemas";
+
+/**
+ * Check if the bucket is valid from the metadata.
+ */
+export function checkIsBucketValidFromMetadata(metadata: BuildMetadata | null) {
+  return !metadata?.testReport || metadata.testReport.status === "passed";
+}
+
+function add(a: number | undefined, b: number | undefined) {
+  if (!a) {
+    return b;
+  }
+  if (!b) {
+    return a;
+  }
+  return a + b;
+}
+
+/**
+ * Aggregate metadata from multiple shards.
+ */
+function aggregateMetadata(allMetatada: (BuildMetadata | null)[]) {
+  return allMetatada.reduce<BuildMetadata | null>((acc, item) => {
+    if (!item?.testReport) {
+      return acc;
+    }
+
+    if (!acc) {
+      return item;
+    }
+
+    const status = (() => {
+      if (acc.testReport?.status === "failed") {
+        return "failed";
+      }
+      if (item.testReport.status === "passed") {
+        return "passed";
+      }
+      return "failed";
+    })();
+
+    const startTime = (() => {
+      if (!acc.testReport?.stats?.startTime) {
+        return item.testReport?.stats?.startTime;
+      }
+      if (!item.testReport?.stats?.startTime) {
+        return acc.testReport.stats.startTime;
+      }
+      return item.testReport.stats.startTime < acc.testReport.stats.startTime
+        ? item.testReport.stats.startTime
+        : acc.testReport.stats.startTime;
+    })();
```

Comment:
```suggestion
  return allMetatada.reduce<BuildMetadata | null>((acc, item) => {
    if (!item?.testReport) return acc;
    if (!acc) return item; 
    
    const status = acc.testReport.status === "failed" || item.testReport.status !== "passed"
      ? "failed"
      : "passed";

    const startTime = (() => {
      const accStartTime = acc.testReport?.stats?.startTime;
      const itemStartTime = item.testReport?.stats?.startTime;
      
      if (!accStartTime) return itemStartTime;
      if (!itemStartTime) return accStartTime;
      
      return Math.min(accStartTime, itemStartTime);
    })();
```

Reply:
It's 100% equivalent right?

Reply:
Yes, it’s a style improvement suggestion.